### DISCUSSION TITLE
Sonoff NSPanel Bluetooth

### DISCUSSION THREAD

**Discussion by Mark-Muc** [arendst/Tasmota#17168] [Q&A]
Hello .. might be someone has an answer.
I compiled an version with online Tasmota-compiler fpr the NSPanel using this defines:
//Sonoff NSPanel Touch Display
#define FIRMWARE_NSPANEL
#define FRAMEWORK_ARDUINO_ITEAD

```
//Thermostat
#define USE_THERMOSTAT                           // Add support for Thermostat
    #define THERMOSTAT_CONTROLLER_OUTPUTS         1         // Number of outputs to be controlled independently
//    #define THERMOSTAT_SENSOR_NAME                "DS18B20" // Name of the local sensor to be used <-- check MQTT!! "ANALOG"
//   MQTT = "ANALOG": {"Temperature1": 20.6}
    #define THERMOSTAT_SENSOR_NAME                "ANALOG" //internal name 
.
.
.
```
also checked:
![image](https://user-images.githubusercontent.com/67410847/203522127-9fa39ba9-1770-4401-801a-ad0bdfa35f82.png)

if I installed this version on the NSPanel .. using this template:
`{"NAME":"NSPanel","GPIO":[0,0,0,0,3872,0,0,0,0,0,32,0,0,0,0,225,0,480,224,1,0,0,0,33,0,0,0,0,0,0,0,0,0,0,4736,0],"FLAG":0,"BASE":1,"CMND":"ADCParam1 2,11200,10000,3950 | Sleep 0 | BuzzerPWM 1"}
`

All worked fine .. did all the stuff described in here:
[](https://templates.blakadder.com/sonoff_NSPanel.html)

it come up ..relais are working ... 
but Blootooth report's strange temperatures:
![image](https://user-images.githubusercontent.com/67410847/203523043-f8eea1e1-b701-4c35-b4c4-780527647dc0.png)

Someone an idea what's happens? I will use an external Temp-Sensor because can get the internal not working.
https://github.com/arendst/Tasmota/discussions/17099#discussioncomment-4169173

and can someone tell me how to set the current temp & gradient temp with MQTT. I looked around an in the documentation 
https://tasmota.github.io/docs/Thermostat/#local-temperature-sensor
i was not able to find .. how to set this 2 fields.

Regards 
Mark-Muc

**Comment by barbudor** [comment]
Do you receive those temperatures exactly at the same time on both ESP32 ?
That sounds weird.
I see signal is very low (-92dB) on the left one
When I'm running 2 ESP32 dev board (one is my prod unit and the other is the dev board for testing) at the same time, I never faced such discrepancy

Regarding other question:
- Current temperature is command [`TempMeasuredSet`](https://tasmota.github.io/docs/Commands/#thermostat)  (you must have `SensorInputSet  0`)
- Gradient cannot be set, it's a measured value. See `TempMeasuredGrdRead`

**Reply by Mark-Muc** [reply]
@barbudor .. so I built a new version with online tasmota compiler .. using:
```
//Sonoff NSPanel Touch Display
#define FIRMWARE_NSPANEL
#define FRAMEWORK_ARDUINO_ITEAD

//Thermostat
#define USE_THERMOSTAT                           // Add support for Thermostat
    #define THERMOSTAT_CONTROLLER_OUTPUTS         1         // Number of outputs to be controlled independently
//    #define THERMOSTAT_SENSOR_NAME                "DS18B20" // Name of the local sensor to be used <-- check MQTT!! "ANALOG"
//   MQTT = "ANALOG": {"Temperature1": 20.6}
    #define THERMOSTAT_SENSOR_NAME                "ANALOG" //internal name 

    #define THERMOSTAT_SENSOR_NUMBER              1         // Number of local sensors to be used
    #define THERMOSTAT_RELAY_NUMBER               1         // Default output relay number for the first controller (+i for following ones)
    #define THERMOSTAT_SWITCH_NUMBER              1         // Default input switch number for the first controller (+i for following ones)
    #define THERMOSTAT_TIME_ALLOW_RAMPUP          300       // Default time after last target update to allow ramp-up controller phase in minutes
    #define THERMOSTAT_TIME_RAMPUP_MAX            960       // Default time maximum ramp-up controller duration in minutes
    #define THERMOSTAT_TIME_RAMPUP_CYCLE          30        // Default time ramp-up cycle in minutes
    #define THERMOSTAT_TIME_SENS_LOST             30        // Maximum time w/o sensor update to set it as lost in minutes
    #define THERMOSTAT_TEMP_SENS_NUMBER           1         // Default temperature sensor number
    #define THERMOSTAT_TIME_MANUAL_TO_AUTO        60        // Default time without input switch active to change from manual to automatic in minutes
    #define THERMOSTAT_TIME_RESET                 12000     // Default reset time of the PI controller in seconds
    #define THERMOSTAT_TIME_PI_CYCLE              30        // Default cycle time for the thermostat controller in minutes
    #define THERMOSTAT_TIME_MAX_ACTION            20        // Default maximum thermostat time per cycle in minutes
    #define THERMOSTAT_TIME_MIN_ACTION            4         // Default minimum thermostat time per cycle in minutes
    #define THERMOSTAT_TIME_MIN_TURNOFF_ACTION    3         // Default minimum turnoff time in minutes, below it the thermostat will be held on
    #define THERMOSTAT_PROP_BAND                  4         // Default proportional band of the PI controller in degrees celsius
    #define THERMOSTAT_TEMP_RESET_ANTI_WINDUP     8         // Default range where reset antiwindup is disabled, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_HYSTERESIS            1         // Default range hysteresis for temperature PI controller, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_FROST_PROTECT         40        // Default minimum temperature for frost protection, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_RAMPUP_DELTA_IN       4         // Default minimum delta temperature to target to get into rampup mode, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_RAMPUP_DELTA_OUT      2         // Default minimum delta temperature to target to get out of the rampup mode, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_PI_RAMPUP_ACC_E       200       // Default accumulated error when switching from ramp-up controller to PI in hundredths of degrees celsius
    #define THERMOSTAT_TIME_OUTPUT_DELAY          180       // Default output delay between state change and real actuation event (f.i. valve open/closed)
    #define THERMOSTAT_TEMP_INIT                  180       // Default init target temperature for the thermostat controller
    #define THERMOSTAT_TIME_MAX_OUTPUT_INCONSIST  3         // Default maximum time where the input and the output shall differ (for diagnostic) in minutes
    #define THERMOSTAT_TIME_MAX_AUTOTUNE          21600     // Maximum time for the PI autotune function to complete in seconds
    #define THERMOSTAT_DUTYCYCLE_AUTOTUNE         35        // Default duty cycle (in % over PI cycle time) for the step response of the autotune PI function
    #define THERMOSTAT_PEAKNUMBER_AUTOTUNE        8         // Default number of peak temperatures (max or min) to be used for the autotune PI function
    #define THERMOSTAT_TEMP_BAND_NO_PEAK_DET      1         // Default temperature band in thenths of degrees celsius within no peak will be detected
    #define THERMOSTAT_TIME_STD_DEV_PEAK_DET_OK   10        // Default standard deviation in minutes of the oscillation periods within the peak detection is successful    
```
flashed and BlueTooth .. flashed it new and TRARA .. working and showing the right Temp on BlueTooth devices!!
So far so good .. BUT i wasn't able to get the internal tempsensor (ANALOG) working.
`SensorInputSet 1`
The temp always shows 0°C.
MQTT tells me: 
`tele/Thermo_WohnZi/SENSOR = {"Time":"2022-11-24T10:01:38","ANALOG":{"Temperature1":25.0},"Thermostat0":{"ThermostatModeSet":1,"TempTargetSet":21.00,"CtrDutyCycleRead":144},"TempUnit":"C"}`
{"Temperature1":25.0} i guess is wrong!!
Should read as: {"Temperature":25.0} (without the "1" at the end ... 
any chance to change?
https://github.com/arendst/Tasmota/pull/8659
![image](https://user-images.githubusercontent.com/67410847/203744162-7a54e775-4c00-4300-b4c2-537716d27fe9.png)
Thx in advance..
Mark-Muc

**Reply by barbudor** [reply]
I will double check but there might be a reason
You can still use a rule 'on analog#temperature1 do...' to set the temperature

**Reply by Mark-Muc** [reply]
Ok Thx .. I'll try ,, the best will be if I can use a BlueTooth Thermometer direct as input.
MQTT Message is:
`10:06:30.549 MQT: tele/Thermo_WohnZi/SENSOR = {"Time":"2022-11-24T10:06:30","ATCaf4a0b":{"alias":"Geraet_1","mac":"a4c138af4a0b","Temperature":21.3,"Humidity":44.0,"DewPoint":8.5,"Battery":61,"RSSI":-36},"LYWSD02116c66":{"alias":"1.OG_Schlafzimmer","mac":"150311116c66","Temperature":16.3,"Humidity":49.0,"DewPoint":5.5,"Battery":26,"RSSI":-103},"TempUnit":"C"}
`
so if i compile a version with internal: ,`"ATCaf4a0b"` i should use the Blotooththerm as "internal" sensor . Should it work like this?
I'll try .. this will be the best methode!!
Regards 
Mark-Muc

**Reply by barbudor** [reply]
I can confirm that on ESP32, as you can have multiple ADC, the index is appended to each measurment.
So for example if you have ADC1 as temperature and ADC2 as Illuminance, the JSON would be 
```JSON
"ANALOG": { "Temperature1": 23.4, "Illuminance2":1000 }
```
So on an ESP32 you should use a rule to trigger on `Temperature1` and  set the temperature with `TempMeasuredSet`

For your bluetooth question, if you have the Mi32Option set to have such SENSOR message, yes setting the internal sensor to the name of the bluetooth device should work.

Still, in any case, you can workaround that type of problems with a rule and `TempMeasuredSet`

**Reply by Mark-Muc** [reply]
@barbudor .. Again thank u for helping so much.
So i built a version of tasmota using: 
`     #define THERMOSTAT_SENSOR_NAME                "ATCaf4a0b" //BlueTooth ID ATCaf4a0b`
the SENSOR message in console:
`11:34:14.398 MQT: tele/Thermo_WohnZi/SENSOR = {"Time":"2022-11-25T11:34:14","ATCaf4a0b":{"alias":"Geraet_1","mac":"a4c138af4a0b","Temperature":21.1,"Humidity":41.0,"DewPoint":7.3,"Battery":62,"RSSI":-44},"LYWSD02116c66":{"alias":"1.OG_Schlafzimmer","mac":"150311116c66","Temperature":16.7,"Humidity":46.0,"DewPoint":5.0,"Battery":26,"RSSI":-103},"TempUnit":"C"}`
so using the name:       "ATCaf4a0b"  should work .. 
`11:42:14.153 MQT: stat/Thermo_WohnZi/RESULT = {"SensorInputSet1":1}`
for my projekt it would be the best solution to have the Current Temperature set in the thermostate.
But so far no luck ...
Current Temperarure shows 0°C ...
Hmm is there any way to debug, to see whats going wrong?

If i have to use Rule:
ON ATCaf4a0b#Temperature DO  TempMeasuredSet <how to get the temp??> ENDON

Thx for any help
Mark-Muc

**Reply by sfromis** [reply]
`TempMeasuredSet %value%`

**Reply by Mark-Muc** [reply]
> `TempMeasuredSet %value%`

Thx ..  work like cham..
Again THX
Mark-Muc
